### ISSUE TITLE
API call failed: Unable to resolve host

### ISSUE THREAD

**Issue by sikoried** [intercom/intercom-android#186]
## Version info
  - Intercom for Android version: 1.1.16
  - Android Device and OS version: Nexus 5 (6.0.1), Samsung S5 (5.0) Samsung S4 (4.4.2), Samsung S3 (4.4.2), all stock.

## Expected behavior
API should be able to connect to Intercom servers.

## Actual behavior
```
E/Intercom-Android: ERROR: Api call failed: Unable to resolve host "ed5qahfg-android-2-0-8-3-0-3-0-8-8-4-2-2-0-1-6-0-3-1-0-1-1-2-2-.mobile-sdk-api.intercom.io": No address associated with hostname
```

## Steps to reproduce
 1. Build app
 2. Launch app
 3. Error will occur right on startup.

We have experienced this since we've upgraded our main app dependencies to retrofit2 and okhttp3.  We're including Intercom as an AAR (as opposed to copying all the code to our repo and building).

Any thoughts?

**Comment by harkin** [comment]
That looks like an `UnknownHostException`. Have to ask to be sure, are you requesting the internet permission and actually connected to the internet when you're seeing these errors?

This worked fine for you before updating to okhttp3 and retrofit2? There shouldn't be any conflicts there, Square deliberately namespaced those libraries differently to avoid that and we namespace the SDK dependencies for the same reason. Does it still work if you don't include okhttp3 and retrofit2?

**Comment by harkin** [comment]
Actually thinking more about this I'd be inclined to think it's a network issue on your end. The exception is when you can't find the IP address for a hostname. It looks like that DNS lookup is failing for some reason for you

**Comment by bscanlan** [comment]
```
$ echo ed5qahfg-android-2-0-8-3-0-3-0-8-8-4-2-2-0-1-6-0-3-1-0-1-1-2-2- | wc -c
      64
```

Illegal label length for a DNS record. https://www.ietf.org/rfc/rfc1123.txt specifies labels (i.e. the thing in between dots) can only be 63 chars or less:
```
 The DNS defines domain name syntax very generally -- a
                 string of labels each containing up to 63 8-bit octets,
                 separated by dots, and with a maximum total of 255
                 octets. 
```

**Comment by harkin** [comment]
@bscanlan looks like wc is including a newline there, I'm counting 63

**Comment by bscanlan** [comment]
Henry wisely pointed out that wc was counting the newline :tired_face:

**Comment by sikoried** [comment]
Both 4G and Wifi connectivity are ok, name resolution works (the networking of our app works, also talking to G analytics).  We also upgraded from 1.1.9 to 1.1.16, I'll step down revisions to see when the issue showed up.

Taking out `okhttp` completely will be a bit work since our app is heavily utilizing it...

**Comment by sikoried** [comment]
So it seems that 1.1.9 works, and the hostname resolution problem appears with update to 1.1.10.  I realize that there's both app-id and platform in the hostname, is that maybe related?

**Comment by sikoried** [comment]
Any update on this issue?  We'd love to upgrade to the latest...

**Comment by sikoried** [comment]
This problem still persists in 1.1.17 :-(

**Comment by harkin** [comment]
@sikoried sorry for the delay in getting back to you about this. I have good news! We found what we believe the issue is and it'll be fixed in the next version (coming very soon)

So what happened? In a previous version (~1.1.11) we switched from having a static URL that every version of this library would use to get in touch with our API to having a dynamic section of that URL, the `ed5qahfg-android-2-0-8-3-0-3-0-8-8-4-2-2-0-1-6-0-3-1-0-1-1-2-2-` part mentioned above. That's made from a few pieces of data and we do some bounds checking to ensure it isn't larger than 63 characters and if it is we'll truncate it. 

The trailing `-` in the first label is non-compliant with [rfc1123](https://www.ietf.org/rfc/rfc1123.txt) so could be rejected by some nameservers. [For example](https://www.whatsmydns.net/#A/ed5qahfg-android-2-0-8-3-0-3-0-8-8-4-2-2-0-1-6-0-3-1-0-1-1-2-2-.mobile-sdk-api.intercom.io)

The solution is pretty simple: ensure there isn't a trailing `-`. It took a long while to get there though.
